### PR TITLE
first draft of base class refactor

### PR DESCRIPTION
**Pull request by frosty00**
This is pretty hefty refactor. I took into account the things you suggested [here.](https://github.com/ccxt/ccxt/pull/1814#issuecomment-363827160)

The code speaks for itself; it splits loadMarkets into `load`, `loadPublic`, `loadPrivate`, `loadMarkets`, `loadCurrencies`, and `loadFees`. It is more extensible than the current base class. Any number of things can be loaded with a simple call to `load()` and if someone wanted to load something more specific they could any one of the lower down loading functions. It is working really well so far and the `fee_loaded` boolean is super helpful for people (like me) who are adverse to using hardcoded fees. 

--- 

Three new methods are added: `load_fees`, `set_fees`, and `propagate_fees`.

* `load_fees`: calls `fetch_fees` --> `set_fees`

* `set_fees`: `self.fees` --> `markets` and `currencies`

* `propagate_fees`: `markets` and `currencies` --> `self.fees`

---

You may want to play around with this code in a python IDE to check for bugs etc. I tested it with a subset of the exchanges offered by ccxt until I could not find anymore bugs. Is it ok that this is a non-backwards compatible change?

Are maker and taker fees allowed to be negative? hitbtc2 defines a negative maker value and I think we should keep the modulus of the trading fees instead - `abs()` the value in python. 

There are likely going to be more changes to this refactor in the future. Could you possibly make a branch so we can test it out more?

### CODE REVIEW

**Comment by kroitor** [comment]
> Are maker and taker fees allowed to be negative?

Yes. This is called a rebate, many exchanges reward you for maker orders, so you don't pay the fee, instead, the exchange pays a part of the fee to you for market-making.

> I think we should keep the modulus of the trading fees instead - abs() the value in python.

No, I don't really think we should lose the sign info. It indicates the direction of the fee.

> Could you possibly make a branch so we can test it out more?

Sure, like I said earlier, will do that shortly, we just wanted to get some testing coverage for the base code first... We have to rework the error handling codepaths in the fetch method. These changes have to be done in specific order, otherwise the branches will quickly run out of sync and it will be very difficult to merge them if they split early.

**Comment by frosty00** [comment]
> We have to rework the error handling codepaths in the fetch method.

What exactly do you mean by this?

Is there any way I can help with refactoring the test cases?

**Comment by frosty00** [comment]
I'm getting a 

    <class 'FileNotFoundError'> [Errno 2] No such file or directory: './keys.json'

**Comment by kroitor** [comment]
> What exactly do you mean by this?

Well, currently it is very convoluted and we have no way of reaching the JSON body in case of a HTTP error. So, we need to rework the error handling there and pass the decoded JSON (if any) to the handle_errors as the last argument (after the raw body). However, this should be done consistently with Py2/3 sync/async. Another thing here is to avoid parsing JSON twice.

> Is there any way I can help with refactoring the test cases?

Your help is always welcome and always appreciated, thanks a lot! We could start by doing some simplification of raise_error and handle_rest_errors... Let me think of it a little, and I'll come up with a proposal in code or here. Thx again!

**Comment by frosty00** [comment]
Another thing we will have to do if we want to refactor `load_markets` is to change `self.load_markets` in other functions to the appropriate method. For example [here](https://github.com/ccxt/ccxt/blob/master/js/bittrex.js#L612) `loadMarkets` is called when really it should be calling `loadCurrencies`. 

Could you elaborate on how and when `raise_error` and `handle_rest_errors` are called?

**Comment by frosty00** [comment]
@kroitor 

Could you explain briefly how errors should be raised in fetch and I'll work on making a PR tomorrow?

Cheers,

**Comment by kroitor** [comment]
@frosty00 yep, will post a sequence here shortly, thx again for your help and for you patience! I'm a little overloaded recently, but I'll get back to you on this very soon.

**Comment by frosty00** [comment]
I see haha - great minds think alike...

**Comment by frosty00** [comment]
I'm looking into getting some sort of consistency with regards to the errors raised. Currently they range from - `print(error, type(error))`:

    livecoin GET https://api.livecoin.net/exchange/order_book?currencyPair=XEM%2FETH&groupByPrice=false request timeout <class 'ccxt.base.errors.RequestTimeout'>
    
    livecoin {"success":false,"errorMessage":"Too Many Requests","errorCode":429} <class 'ccxt.base.errors.ExchangeError'>

    [Errno 54] Connection reset by peer <class 'aiohttp.client_exceptions.ClientOSError'> 

    bitfinex {"error":"ERR_RATE_LIMIT"} <class 'ccxt.base.errors.DDoSProtection'>

I think we should try unify all the errors so they look like 2 and 4 - (1 and 3 if you like zero-indexing). 

---

Ideally when `str` or `repr` is called on our errors they should look the same or at least homologous. I propose that all errors start with `self.id` and then give some information about the error. A json response is sufficient. Could you explain the order in which errors are should be processed starting from fetch.

**Comment by kroitor** [comment]
> I'm looking into getting some sort of consistency with regards to the errors raised. 

Yep, this is something we need as well, for sure.

> `[Errno 54] Connection reset by peer <class 'aiohttp.client_exceptions.ClientOSError'> `

This particular one is missing a handler in the `python/ccxt/async/base/exchange.py`. We should convert it to ExchangeError or ExchangeNotAvailable.

> Could you explain the order in which errors should be processed starting from fetch?

1. get response headers and body
2. json-parse it just once at most
3. call the unified error handler first – `handle_errors()`from the base class does nothing mostly... it expects headers and response body along with the error code in its arguments.
4. if it passed the unified error handler with no exceptions, call the additional base handler for any errors that were missed in case handle_errors isn't defined on a derived exchange... the base handler catches HTTP errors and some Cloudflare protection errors...
5. if it passed both handlers, then the response handler is called

Let me know if this answers the question )

**Comment by frosty00** [comment]
Instead of

    livecoin GET https://api.livecoin.net/exchange/order_book?currencyPair=XEM%2FETH&groupByPrice=false request timeout <class 'ccxt.base.errors.RequestTimeout'>

would it be possible to have 'exchange_name' as the first argument then a dictionary as the second argument, like so

    livecoin {'url': 'https://api.livecoin.net/exchange/order_book?currencyPair=XEM%2FETH&groupByPrice=false'} <class 'ccxt.base.errors.RequestTimeout'>

I think most errors could use this format. Thus the exchange which raised the error is always:

    try:
        bittrex.fetch_order_book()
    except ccxt.BaseError as error:
        print(error.args[0]) # exchange name
        print(error.args[1]) # dictionary with details relating to the error

**Comment by kroitor** [comment]
> What are your thoughts?

We were actually thinking of extending the BaseError class in all three languages with the following fields: 

```
{
  'exchange': 'foobar',
  'type': 'AuthenticationError',
  'message': 'bad signature',
  'http_status_code': 403,
  'url': 'https://api.....com/balance',
  'method': 'fetchBalance',
  'request': { 'headers': .., 'body': .. },
  'response': { 'headers': ..., 'body': .. },
  'file': '...',
  'line': 123,
  'stacktrace': ...
}
```

In that case all errors would be standardized:

```
try:
    bittrex.fetch_order_book()
except ccxt.BaseError as error:
    print(error.exchange) # exchange name
    print(error.message) # dictionary with details relating to the error
```

 Some fields may be redundant, some other may be missing and need to be added there.... However, unifying it across all languages and calls may be harder than it seems... %)

**Comment by frosty00** [comment]
I can't make the load_fees asynchronous because `fetchFundingFees` and `fetchTradingFees` are not defined as aschronous (python) methods in subclasses. I can't think of a way to make them work asynchronously and sychronously. Maybe there is a way to wrap regular functions as futures with asyncio?

**Comment by frosty00** [comment]
Also would it be possible to transpile `load_markets` and `load_currencies` from the asynchronous version to the non-asynchronous version so that they stay the same when someone makes changes to one of them?

**Comment by kroitor** [comment]
Sorry, I'm late with the answers for your most recent questions, but, I see that you have figured it out yourself )

**Comment by frosty00** [comment]
In the case that a market specifies different fees for different pairs what should we set `self.fees.trading` equal to. Currently it only contains maker and taker fees, and I have avoided setting it in the case that the market's fees are not all the same. 

However, maybe it would be wiser to do something simillar to `self.fees.funding` whereby each currency has it's own fee specified, but with markets instead. 

i.e. 
Currently:

    {'fees':
        'trading': {
            'maker': 0.01,
            'taker': 0.004,
        }
        'funding': {
            'XRP': 2,
            'ETH: 12,
            'XYZ': ...
        }
    }

---

Instead it could be:

    {'fees':
        'trading': {
            'BTC/ETH': {
                        'maker': 0.001,
                        'taker': 0.001,
                                }
            'LYM/BTC': {
                        'maker': 0.0001,
                        'taker': 0.00123,
                                }
                        }
                ...
    }

**Comment by kroitor** [comment]
@frosty00 this is exactly how we would want it, per-market )

**Comment by frosty00** [comment]
Could you please look over this code for any bugs. From what I have tested it with it is working really well and would be a worthwhile addition to the library. Of course first we have to refactor all the derived exchange classes in javascript...

**Comment by kroitor** [comment]
@frosty00 i'm looking over all your commits, you're doing a great job, really! The reason why it is not merged yet, is that I'm having difficulties finding time to finish the same work that I do for JS and PHP, I'll update you and will merge this PR asap, most likely within a day or a couple of days. Thanks and respect!

**Comment by kroitor** [comment]
One consideration that I have with public=False, private=False, markets=False, currencies=False, and so on – with listing predefined positional args in general – we already have issues with this type of design... Namely, with createOrder we occasionally find (about once in a quarter) another argument that has strong reasons to be added to the list of arguments, therefore we think of switching to dict-based arguments. So it might turn into `def load(self, options={}):`.

However, this will be less arg-count-and-type-safe, because it would be easier to miss an argument or to make a typo in its name, if instead of being a method argument, it will be a value in a dictionary under a string-key. With python this is a less critical issue because in most cases you can use positional and kw-args interchangeably in python... But it's a little more restricted in other languages, thus, it looks like we have to make a very difficult decision. Let me know what you think.

**Comment by mkutny** [comment]
> However, this will be less arg-count-and-type-safe, because it would be easier to miss an argument or to make a typo in its name

I once happened to `createOrder (symbol, amount, price)`. But I'm not alone, just two recent cases when users mixed the order of args too: #2040, #2044.

We, humans, are really good at making mistakes. We've been doing mistakes with positional args and we **will** be doing mistakes with options pattern too. Still, both are valid patterns and options pattern is no more error-prone than positional args one.

The bottom line: I'd put inevitable user mistakes aside when we choose one over the other.

BTW, we use options pattern for exchange instantiation and... it seems that users are happy about it )

**Comment by kroitor** [comment]
@mkutny looks like I need to come up with some plan on how we are going to actually switch to the *options pattern*...

**Comment by frosty00** [comment]
Options by itself would not be particularly userfriendly, we would need to use a docstring of some sort to provide visual feedback when they type the arguments or call `help()` on the function. Also I don't think having a dictionary provides any advantages over just using keyword arguments (in python at least).

**Comment by kroitor** [comment]
> Options by itself would not be particularly userfriendly

We will lose intellisense, indeed.

> in python if you want to convert a dictionary to keyword args you can do a `**`

Yep, decomposing arg lists and assignments are also available in JS, and, most likely, in PHP as well... However, they have very different semantics, thus, not all trivial to transpile properly...

This question is related to https://github.com/ccxt/ccxt/pull/1894 – adding a support for timeInForce to createOrder... There are two options: add another backward-incompatible argument or convert to options entirely to be less "breaking" upon later changes...

**Comment by frosty00** [comment]
Something that we should consider when using mutable default arguments is their weird behaviour in some languages. 

Take for example:

    def load(options={}):
        print(options)
        options['x'] = True

---

    >>>load()
    {}
    >>>load()
    {'x': True}

**Comment by kroitor** [comment]
@frosty00 very interesting... %)
